federated_learning/run_toxicity.py

```diff
@@ -0,0 +1,223 @@
+    h = layers.GraphGather()(h)
+    logits = tf.keras.layers.Dense(1, activation='sigmoid')(h)
+    return tf.keras.Model(inputs=[input_adjs, input_features], outputs=logits)
+
+
+def build_model_gcn(max_n_atoms, max_n_types):
+    input_adjs = tf.keras.Input(
+        shape=(1, max_n_atoms, max_n_atoms), name='adjs', sparse=False)
+    input_features = tf.keras.Input(
+        shape=(max_n_atoms, max_n_types), name='features')
+    # for graph
+    h = layers.GraphConvFL(64, 1)(input_features, input_adjs)
+    h = tf.keras.layers.ReLU()(h)
+    h = layers.GraphConvFL(64, 1)(h, input_adjs)
+    h = tf.keras.layers.ReLU()(h)
+    h = layers.GraphGather()(h)
+    logits = tf.keras.layers.Dense(1, activation='sigmoid', name="dense")(h)
+    return tf.keras.Model(inputs=[input_adjs, input_features], outputs=logits)
+
+
+def client_data(source, n, batch_size, epochs):
+    return source.create_tf_dataset_for_client(source.client_ids[n]).repeat(epochs).batch(batch_size)
+
+
+@click.command()
+@click.option('--federated', is_flag=True)
+@click.option('--rounds', default=10, help='the number of updates of the centeral model')
+@click.option('--clients', default=2, help='the number of clients')
+@click.option('--epochs', default=10, help='the number of training epochs in client traning.')
+@click.option('--batchsize', default=32, help='the number of batch size.')
+@click.option('--lr', default=0.2, help='learning rate for the central model.')
+@click.option('--clientlr', default=0.001, help='learning rate for client models.')
+@click.option('--model', default='gcn', type=click.Choice(['gcn', 'gin']), help='support gcn or gin.')
+@click.option('--ratio', default=None, help='set ratio of the biggest dataset in total datasize.' +
+              ' Other datasets are equally divided. (0, 1)')
+@click.option('--dataset_name', type=click.Choice(['Benchmark', 'NTP_PubChem_Bench.20201106', 'Ames_S9_minus.20201106']),
+              default='NTP_PubChem_Bench.20201106', help='set dataset name')
+def main(federated, rounds, clients, epochs, batchsize, lr, clientlr, model, ratio, dataset_name):
+    logger = get_logger()
+    subsets = clients + 2
+    logger.debug(f'federated = {federated}')
+    logger.debug(f'rounds = {rounds}')
+    logger.debug(f'clients = {clients}')
+    logger.debug(f'subsets = {subsets}')
+    logger.debug(f'epochs = {epochs}')
+    logger.debug(f'batchsize = {batchsize}')
+    logger.debug(f'lr = {lr}')
+    logger.debug(f'clientlr = {clientlr}')
+    logger.debug(f'model = {model}')
+    logger.debug(f'ratio = {ratio}')
+    logger.debug(f'dataset_name = {dataset_name}')
+
+    if federated:
+        federated_learning(rounds, clients, epochs, batchsize,
+                           lr, clientlr, model, ratio, dataset_name)
+    else:
+        normal_learning(rounds, epochs, batchsize, lr, model, dataset_name)
+
+
+def federated_learning(rounds, clients, epochs, batchsize, lr, clientlr, model, ratio, dataset_name):
```

Comment:
This is also the imitation of `run_chembl.py`.  This ran with the following log. 
```
2020-12-18 11:02:13,197 - run_toxicity.py - main - L 65 ::: rounds = 10
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 66 ::: clients = 2
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 67 ::: subsets = 4
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 68 ::: epochs = 10
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 69 ::: batchsize = 32
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 70 ::: lr = 0.2
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 71 ::: clientlr = 0.001
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 72 ::: model = gcn
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 73 ::: ratio = None
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 74 ::: task = NTP_PubChem_Bench.20201106
2020-12-18 11:02:13,198 - run_toxicity.py - main - L 83 ::: ratios = None
2020-12-18 11:02:23,520 - run_toxicity.py - main - L 86 ::: ['TOX000', 'TOX001', 'TOX002', 'TOX003']
2020-12-18 11:02:23,520 - run_toxicity.py - main - L 87 ::: 4
2020-12-18 11:02:24.014340: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-12-18 11:02:24.035132: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fc5e602a5c0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-12-18 11:02:24.035149: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:tensorflow:From /Users/yuta.kohno/.pyenv/versions/miniconda3-latest/envs/tf2/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:1666: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
2020-12-18 11:02:31,407 - run_toxicity.py - main - L126 ::: 0 round ->
2020-12-18 11:03:17,058 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 000, 1.45478, 0.42624, 0.41665,
2020-12-18 11:03:46,205 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 000, 6.16335, 0.47436, 0.48317,
2020-12-18 11:04:36,532 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 001, 1.62242, 0.37068, 0.33218,
2020-12-18 11:04:55,587 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 001, 2.27475, 0.50588, 0.47495,
2020-12-18 11:05:34,155 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 002, 1.33292, 0.39628, 0.38714,
2020-12-18 11:05:50,502 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 002, 1.42018, 0.54487, 0.59081,
2020-12-18 11:06:27,208 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 003, 1.11567, 0.40969, 0.40584,
2020-12-18 11:06:43,540 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 003, 1.93251, 0.49679, 0.49747,
2020-12-18 11:07:20,625 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 004, 0.97015, 0.41368, 0.40186,
2020-12-18 11:07:36,969 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 004, 0.88452, 0.50267, 0.51775,
2020-12-18 11:08:14,918 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 005, 0.70589, 0.50756, 0.54124,
2020-12-18 11:08:31,755 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 005, 1.07060, 0.52831, 0.56019,
2020-12-18 11:09:09,942 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 006, 0.75871, 0.56104, 0.58089,
2020-12-18 11:09:26,924 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 006, 1.20966, 0.53365, 0.57419,
2020-12-18 11:10:05,600 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 007, 0.79432, 0.58261, 0.60499,
2020-12-18 11:10:23,233 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 007, 0.81901, 0.56303, 0.59968,
2020-12-18 11:11:01,822 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 008, 0.72368, 0.57597, 0.59332,
2020-12-18 11:11:17,585 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 008, 0.75213, 0.57853, 0.60402,
2020-12-18 11:11:55,850 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 009, 0.69453, 0.56798, 0.58139,
2020-12-18 11:12:12,965 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 009, 0.72555, 0.51923, 0.59598,
2020-12-18 11:12:31,398 - run_toxicity.py - main - L157 ::: test, round, loss, acc, auc ===> 0, 009, 0.71788,  0.51131, 0.58524,
2020-12-18 11:12:31,522 - run_toxicity.py - main - L126 ::: 1 round ->
2020-12-18 11:13:11,473 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 000, 1.59658, 0.40891, 0.41474,
2020-12-18 11:13:28,903 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 000, 1.79132, 0.46886, 0.45156,
2020-12-18 11:14:07,933 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 001, 0.75220, 0.44898, 0.41786,
2020-12-18 11:14:25,234 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 001, 0.72293, 0.46923, 0.50452,
2020-12-18 11:15:05,042 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 002, 0.72409, 0.44264, 0.40541,
2020-12-18 11:15:22,720 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 002, 0.88136, 0.54241, 0.55131,
2020-12-18 11:16:03,470 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 003, 0.82290, 0.39816, 0.38508,
2020-12-18 11:16:20,459 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 003, 0.77909, 0.48353, 0.50844,
2020-12-18 11:17:02,793 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 004, 0.71339, 0.53043, 0.44991,
2020-12-18 11:17:20,171 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 004, 0.75755, 0.51747, 0.53031,
2020-12-18 11:17:57,589 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 005, 0.70443, 0.54420, 0.51400,
2020-12-18 11:18:14,786 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 005, 0.73021, 0.54454, 0.59905,
2020-12-18 11:18:52,183 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 006, 0.71432, 0.56037, 0.58301,
2020-12-18 11:19:08,485 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 006, 0.85261, 0.54077, 0.62210,
2020-12-18 11:19:44,885 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 007, 0.74438, 0.56426, 0.56317,
2020-12-18 11:20:01,319 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 007, 0.92924, 0.55583, 0.60182,
2020-12-18 11:20:42,336 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 008, 0.74565, 0.54908, 0.57097,
2020-12-18 11:20:59,283 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 008, 1.25514, 0.51192, 0.55441,
2020-12-18 11:21:39,161 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 009, 0.70313, 0.56599, 0.58553,
2020-12-18 11:21:56,293 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 009, 0.75256, 0.55530, 0.59841,
2020-12-18 11:22:14,195 - run_toxicity.py - main - L157 ::: test, round, loss, acc, auc ===> 1, 009, 0.74669,  0.55288, 0.59383,
2020-12-18 11:22:14,316 - run_toxicity.py - main - L126 ::: 2 round ->
2020-12-18 11:22:54,438 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 000, 1.41415, 0.42894, 0.42925,
2020-12-18 11:23:09,907 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 000, 0.78209, 0.43595, 0.42666,
2020-12-18 11:23:49,067 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 001, 0.71123, 0.52886, 0.44749,
2020-12-18 11:24:04,610 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 001, 0.69094, 0.54246, 0.55419,
2020-12-18 11:24:44,247 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 002, 0.69111, 0.53356, 0.39607,
2020-12-18 11:25:02,219 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 002, 4.67062, 0.45382, 0.45587,
2020-12-18 11:25:43,642 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 003, 0.84323, 0.49799, 0.42593,
2020-12-18 11:25:59,940 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 003, 0.68421, 0.55896, 0.52327,
2020-12-18 11:26:39,369 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 004, 0.68774, 0.53965, 0.34511,
2020-12-18 11:26:55,217 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 004, 0.68282, 0.55452, 0.51759,
2020-12-18 11:27:36,641 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 005, 0.68829, 0.53884, 0.41849,
2020-12-18 11:27:54,219 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 005, 0.68703, 0.55507, 0.51836,
2020-12-18 11:28:36,384 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 006, 0.68875, 0.53742, 0.41262,
2020-12-18 11:28:54,264 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 006, 1.44828, 0.47152, 0.48889,
2020-12-18 11:29:36,958 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 007, 0.69160, 0.53471, 0.38912,
2020-12-18 11:29:53,751 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 007, 3.25259, 0.46376, 0.49422,
2020-12-18 11:30:36,082 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 008, 0.71698, 0.52833, 0.35706,
2020-12-18 11:30:53,988 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 008, 0.69174, 0.56118, 0.51650,
2020-12-18 11:31:34,662 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 009, 0.68631, 0.54128, 0.29550,
2020-12-18 11:31:51,286 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 009, 0.68812, 0.56507, 0.52407,
2020-12-18 11:32:08,670 - run_toxicity.py - main - L157 ::: test, round, loss, acc, auc ===> 2, 009, 0.69474,  0.54614, 0.51013,
2020-12-18 11:32:08,798 - run_toxicity.py - main - L126 ::: 3 round ->
2020-12-18 11:32:46,784 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 000, 1.16117, 0.36387, 0.37063,
2020-12-18 11:33:03,058 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 000, 7.22985, 0.46886, 0.49032,
2020-12-18 11:33:44,349 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 001, 1.42428, 0.39305, 0.35148,
2020-12-18 11:34:00,986 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 001, 0.83299, 0.54031, 0.55945,
2020-12-18 11:34:39,692 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 002, 0.76377, 0.51988, 0.52631,
2020-12-18 11:34:55,894 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 002, 0.68725, 0.53969, 0.62985,
2020-12-18 11:35:36,931 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 003, 0.70411, 0.54928, 0.52174,
2020-12-18 11:35:53,969 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 003, 0.98741, 0.47842, 0.49268,
2020-12-18 11:36:34,188 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 004, 0.70931, 0.55115, 0.54298,
2020-12-18 11:36:51,002 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 004, 0.69811, 0.56854, 0.57576,
2020-12-18 11:37:31,048 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 005, 0.70659, 0.55086, 0.54051,
2020-12-18 11:37:48,061 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 005, 3.65569, 0.46616, 0.50593,
2020-12-18 11:38:27,968 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 006, 0.78583, 0.51730, 0.52559,
2020-12-18 11:38:44,847 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 006, 6.34853, 0.46352, 0.47845,
2020-12-18 11:39:24,492 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 007, 1.00269, 0.50021, 0.53682,
2020-12-18 11:39:41,700 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 007, 0.70577, 0.53227, 0.57540,
2020-12-18 11:40:21,093 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 008, 0.68464, 0.54262, 0.56572,
2020-12-18 11:40:38,166 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 008, 2.20804, 0.48583, 0.48875,
2020-12-18 11:41:17,916 - run_toxicity.py - main - L133 ::: train, round, loss, acc, auc  ===> 009, 0.71285, 0.53010, 0.52563,
2020-12-18 11:41:34,143 - run_toxicity.py - main - L139 ::: val, round, loss, acc, auc ===> 009, 3.29861, 0.47786, 0.48148,
2020-12-18 11:41:49,956 - run_toxicity.py - main - L157 ::: test, round, loss, acc, auc ===> 3, 009, 3.53723,  0.47153, 0.47736,
2020-12-18 11:41:49,957 - run_toxicity.py - main - L159 ::: 2 >>>> all_test_acc = [0.51131463, 0.55288464, 0.54613906, 0.4715296]
2020-12-18 11:41:49,957 - run_toxicity.py - main - L160 ::: 2 >>>> all_test_loss = [0.7178811, 0.7466932, 0.69474113, 3.5372334]
2020-12-18 11:41:49,957 - run_toxicity.py - main - L161 ::: 2 >>>> all_test_auc = [0.58523566, 0.59383476, 0.5101271, 0.47735915]
```